File: igmrf/rw2.R
rm(list=ls())
set.seed(123)
library(tidyverse)
library(TMB)
library(ar.matrix)
library(mvtnorm)
library(sparseMVN)
library(Matrix)

simulateRW2 <- function(N, sigma_=1, burnin=1000){
    resultsplus <- rnorm(burnin+N, 0, (sigma_/2))
    for(i in 3:(burnin+N)){
        resultsplus[i] <- resultsplus[i] +
            ((.5)*resultsplus[i-1]) + ((.5)*resultsplus[i-2])
    }
    resultsT <- resultsplus[-1:-burnin]
    results <- resultsT - mean(resultsT)
    return(results)
}


rw2Q <- function(M, sparse=T){
    Q <- Matrix::Matrix(0, nrow = M, ncol = M)
    Q[1, 1:3] <- c(1, -2, 1)
    Q[2, 1:4] <- c(-2, 5, -4, 1)
    for (i in 3:(M-2)){
        Q[i, (i-2):(i+2)] <- c(1, -4, 6, -4, 1)
    }
    Q[M-1, (M-3):M] <- c(1, -4, 5, -2)
    Q[M, (M-2):M] <- c(1, -2, 1)
    if(sparse)
        Q <- Matrix::Matrix(Q, sparse = TRUE)
    Q
}

testDifferences <- function(x, second=T){
    if(second){
        xdiff <- sapply(3:length(x), function(i){
            x[i] - (2 * x[i-1]) + x[i-2]
        })
    }
    else{
        xdiff <- sapply(2:length(x), function(i){
            x[i] - x[i-1]
        })
    }
    xdiff
}

trueValue <- 1.4

x <- simulateRW2(400, trueValue)
plot(x)
plot(density(testDifferences(x)))
quantile(testDifferences(x), probs = c(.025, .975))

densRW2 <- function(log_sigma, x_=x, log=T, alt=F, offset=.00001){
    sigma_ <- exp(log_sigma)
    kappa_ <- 1/(sigma_^2)
    N <- length(x_)
    Q <- rw2Q(N) * kappa_
    nll <- -1 * (((N-2)/2) * log(kappa_) + (-.5 * ((t(x_) %*% Q %*% x_)[1,1])))
    if(!log){
        nll <- exp(nll)
    }
    if(alt){
        diag(Q) <- diag(Q) + .00001
        CH <- Cholesky(Q)
        nll <- -dmvn.sparse(x_, rep(0, nrow(Q)), CH)
    }
    return(nll)
}

modelRun <- function(x, verbose=T){
    model <- "rw2"
    compile(paste0(model, ".cpp"))
    dyn.load(dynlib(model))
    Data <- list(x=x)
    Params <- list(log_sigma=0)
    
    Obj <- MakeADFun(data=Data, parameters=Params, DLL=model, silent=!verbose)
    
    Opt <- nlminb(start=Obj$par, objective=Obj$fn, 
                  gradient=Obj$gr,
                  control=list(eval.max=1e6, iter.max=1e6))
    
    sdrep <- sdreport(Obj, getJointPrecision=TRUE)
    dyn.unload(dynlib(model))
    
    list(obj=Obj, opt=Opt, sd=sdrep)
}


test <- modelRun(x)
test2 <- optim(
    0, densRW2, method="Brent", lower=-100, upper=100, hessian=T, alt=T)


data.frame(
    method=c("tmb", "optim"),
    mu=c(unname(exp(test$sd$par.fixed)), exp(test2$par)),
    lwr=exp(c(
        test$sd$par.fixed - 1.96 * test$sd$cov.fixed^.5,
        test2$par - 1.96 * test2$hessian^-.5)),
    upr=exp(c(
        test$sd$par.fixed + 1.96 * test$sd$cov.fixed^.5,
        test2$par + 1.96 * test2$hessian^-.5)),
    true=c(trueValue, trueValue))

File: igmrf/rw2.cpp
#include <TMB.hpp>
#include <Eigen/Sparse>
#include <vector>
using namespace density;
using Eigen::SparseMatrix;
template<class Type>
SparseMatrix<Type> spde_Q(Type logkappa, Type logtau, SparseMatrix<Type> M0,
                          SparseMatrix<Type> M1, SparseMatrix<Type> M2) {
    SparseMatrix<Type> Q;
    Type kappa2 = exp(2. * logkappa);
    Type kappa4 = kappa2*kappa2;
    Q = pow(exp(logtau), 2.)  * (kappa4*M0 + Type(2.0)*kappa2*M1 + M2);
    return Q;
}

template<class Type>
matrix<Type> rw2_Q(int N, Type sigma) {
    matrix<Type> Q(N,N);
    Q(0,0) = (1.) / pow(sigma, 2.);
    Q(0,1) = (-2.) / pow(sigma, 2.);
    Q(1,0) = (-2.) / pow(sigma, 2.);
    Q(1,1) = (5.) / pow(sigma, 2.);
    for (int n = 2; n < (N-1); n++) {
        Q(n,n) = (6.) / pow(sigma, 2.);
        Q(n-1,n) = (-4.) / pow(sigma, 2.);
        Q(n,n-1) = (-4.) / pow(sigma, 2.);
        Q(n-2,n) = (1.) / pow(sigma, 2.);
        Q(n,n-2) = (1.) / pow(sigma, 2.);
    }
    Q(N-2,N-2) = (5.) / pow(sigma, 2.);
    Q(N-1,N-2) = (-2.) / pow(sigma, 2.);
    Q(N-2,N-1) = (-2.) / pow(sigma, 2.);
    Q(N-1,N-1) = (1.) / pow(sigma, 2.);
    return Q;
}

template<class Type>
SparseMatrix<Type> rw2_QS(int N, Type sigma) {
    Type offset = Type(.0001);
    SparseMatrix<Type> Q(N,N);
    Q.insert(0,0) = (1.) / pow(sigma, 2.) + offset;
    Q.insert(0,1) = (-2.) / pow(sigma, 2.);
    Q.insert(1,0) = (-2.) / pow(sigma, 2.);
    Q.insert(1,1) = (5.) / pow(sigma, 2.) + offset;
    for (int n = 2; n < (N-1); n++) {
        Q.insert(n,n) = (6.) / pow(sigma, 2.) + offset;
        Q.insert(n-1,n) = (-4.) / pow(sigma, 2.);
        Q.insert(n,n-1) = (-4.) / pow(sigma, 2.);
        Q.insert(n-2,n) = (1.) / pow(sigma, 2.);
        Q.insert(n,n-2) = (1.) / pow(sigma, 2.);
    }
    Q.coeffRef(N-2,N-2) = (5.) / pow(sigma, 2.) + offset;
    Q.coeffRef(N-1,N-2) = (-2.) / pow(sigma, 2.);
    Q.coeffRef(N-2,N-1) = (-2.) / pow(sigma, 2.);
    Q.insert(N-1,N-1) = (1.) / pow(sigma, 2.) + offset;
    return Q;
}

template<class Type>
Type densRW2(vector<Type>x, Type sigma) {
    Type kappa_ = pow(sigma, -2.);
    int N = x.size();
    matrix<Type> xt(1, N);
    for(int n = 1; n < (N); n++){
        xt(0, n) = x[n]; 
    }
    matrix<Type> Q = rw2_Q(N, sigma);
    matrix<Type> temp = xt * Q;
    temp = temp * x;
    Type nll = -1. * (((Type(N) - 2.) / 2.) * log(kappa_) + (-.5 * temp(0,0)));
    return nll;
}

template<class Type>
Type objective_function<Type>::operator() ()
{
   
    DATA_VECTOR(x);
    PARAMETER(log_sigma);
    
    Type sigma = exp(log_sigma);
    
    matrix<Type> denseQ = rw2_Q(x.size(), sigma);
    
    for(int i=0;i<x.size();i++) {
        denseQ(i,i) += .00001;
    }
    
    Eigen::SparseMatrix<Type> Q = asSparseMatrix(denseQ);
    
    Type nll = GMRF(Q)(x);
    // Type nll = densRW2(x, sigma);
    
    REPORT(log_sigma);
    return nll;
}
